[build] Fix and extend build scripting

benyamin-codez opened the pull request:
1. Fixed various syntax issues, e.g. unquoted variables
2. Unwrapped nested `IF` statements to ensure no jumping out of loops
3. Extended colour pallette to `stdout`
4. Introduced environment variable `SKIP_SDV_ACTUAL` to skip _Static Driver Verifier (SDV)_ during analysis builds. _CodeQL, Code Analysis (CA)_ and _DVL_ operations are still executed. Default is to perform the SDV build (unchanged behaviour).
5. Make actual SDV build conditional (`Win10_SDV` only) f)
6. Introduced managed CodeQL package cache and test suite versioning control. If the CodeQL binary is detected at `CODEQL_BIN` a CodeQL build commences, otherwise it is skipped (unchanged behaviour). The package cache versions are set in the `:config_ql_whcp` function (new). Test suite versioning is via git hash of the WDK Developer Supplemental Tools repository (new, depends on `git`). The versions are determined by the value of the `WHCP_LEVEL` variable, which can be either `WHCP_LEGACY` or `WHCP_24H2`. Provision for future WHCP versions is templated as `WHCP_NEXT`.
7. Introduced environment variable `CODEQL_OFFLINE_ONLY` to perform package cache and suite version checking but NOT download any needed updates. Exits on error expecting resolution. Requires `git`.
8. Introduced environment variable `CODEQL_RUN_BLIND` to retain legacy behaviour and perform no package cache or suite version checking. It presumes the prerequisites are present. Removes `git` dependency.
9. Pre-builds x86 `viosock` libraries when building `virtio-win.sln` or `viosock.sln` for `amd64` in the circumstances the required x86 libraries do not already exist.
10. If running in a Germanium EWDK environment, the script will instantiate a Cobalt EWDK environment to build Win10 SDV operations, and will return to Germanium EWDK when done.
11. Use Germanium EWDK to sign Win11 drivers.
12. Enforce `echo off` after calling `SetupBuildEnv.cmd`
13. Provide some extra EWDK details when `EnterpriseWDK=True`
14. Prettify last leg of build and provide SUCCESS / FAIL feedback.

kostyanf14 commented:
@benyamin-codez 

1. Please split this into several commits/PRs to simplify the review
2. Several items are really independent and can be merged separately, so create new PRs for it:
   - 1. Fixed various syntax issues, e.g. unquoted variables
   - 2. Unwrapped nested IF statements to ensure no jumping out of loops
   - 11. Use Germanium EWDK to sign Win11 drivers.
   - 12. Enforce echo off after calling SetupBuildEnv.cmd
   - 13. Provide some extra EWDK details when EnterpriseWDK=True
3. Move item `9. Pre-builds x86 viosock libraries when building...` to separate PR for discussion
4. Why do we need `4. Introduced environment variable SKIP_SDV_ACTUAL to skip`? What is the point of building DVL without SDV results? For real certification for Win10/Win11 (before 24H2) you must submit DVL with SDV. If you want to check CodeQL results you can build a Win11_SDV target that skips SDV as it is deprecated. 
5. Why do we need `6. Introduced managed CodeQL package cache and test suite versioning control.`? CodeQL results with the Win11 24H2 test suite are fully compatible with all current submissions that require CodeQL results.

benyamin-codez commented:
@kostyanf14 

Thanks for having a look, Kostiantyn..!

First let me say, I've got some more changes to commit. So I'll do that first and then look at breaking it up.

I'll likely put these two together to avoid merge dependencies:

`1. Fixed various syntax issues, e.g. unquoted variables`
`2. Unwrapped nested IF statements to ensure no jumping out of loops`

I'll probably drop these together as they're in the same file [`build\SetVsEnv.bat`] and minor:

`12. Enforce echo off after calling SetupBuildEnv.cmd`
`13. Provide some extra EWDK details when EnterpriseWDK=True`

> 3. Move item `9. Pre-builds x86 viosock libraries when building...` to separate PR for discussion

Can do.
I must say, I'm quite curious to see what you might have to say about this.
I would imagine most - if not all - of your workflows would never hit this scenario, i.e. building without x86, whereas I would very rarely build for x86. Just a final build including x86 to make sure there are no compilation errors tbh.

> 4. Why do we need `4. Introduced environment variable SKIP_SDV_ACTUAL to skip...`? What is the point of building DVL without SDV results? For real certification for Win10/Win11 (before 24H2) you must submit DVL with SDV. If you want to check CodeQL results you can build a Win11_SDV target that skips SDV as it is deprecated.

One scenario is when testing the build system. Without having to do the actual SDV to check the tail end of the build, this is a great time saver. It's also helpful when you have already done the SDV build and you don't want to clobber it on a re-run. Another scenario is if you want to build for any target and do code analysis but not SDV. I note your mention of using Win11_SDV, but sometimes you might want to build for Win10 and not wish to expose Win11-specific options.

Obviously, these are all prototyping scenarios and not building for the WHCP.

>5. Why do we need `6. Introduced managed CodeQL package cache and test suite versioning control.`? CodeQL results with the Win11 24H2 test suite are fully compatible with all current submissions that require CodeQL results.

That's certainly how we do it - and I'm in favour of that. However, I followed the MS documentation [here](https://learn.microsoft.com/en-us/windows-hardware/drivers/devtest/static-tools-and-codeql#for-windows-hardware-compatibility-program-use) and [here](https://github.com/microsoft/Windows-Driver-Developer-Supplemental-Tools#windows-driver-developer-supplemental-tools) in implementing this _framework_, and that really is what I am providing with this feature, a framework for future versions.
Let me provide an example - here is what this commit introduces:

```batch
if "%WHCP_LEVEL%"=="WHCP_LEGACY" (
  set CODEQL_PACK_DEL=microsoft/windows-drivers@1.1.0
  set CODEQL_PACK_GET=codeql/cpp-queries@0.9.0,microsoft/windows-drivers@1.0.13
)
if "%WHCP_LEVEL%"=="WHCP_24H2" (
  set CODEQL_PACK_DEL=
  set CODEQL_PACK_GET=codeql/cpp-queries@0.9.0,microsoft/windows-drivers@1.1.0
)
if "%WHCP_LEVEL%"=="WHCP_NEXT" (
  rem USE THIS AS A TEMPLATE FOR FUTURE VERSIONS
  rem NOTE: Remember to populate the CODEQL_PACK_DEL variable for all WHCP_LEVELS above with __NEXTRANGE__ packs [per WHCP_LEGACY example]...
  set CODEQL_PACK_DEL=
  set CODEQL_PACK_GET=codeql/cpp-queries@__NEXTRANGE__,microsoft/windows-drivers@__NEXTRANGE__
)
```
...and then, with the introduction of say an imaginary WHCP_25H2 and updated CodeQL packs, we might have:

```batch
if "%WHCP_LEVEL%"=="WHCP_LEGACY" (
  set CODEQL_PACK_DEL=microsoft/windows-drivers@1.1.0,codeql/cpp-queries@0.10.0,microsoft/windows-drivers@1.2.0
  set CODEQL_PACK_GET=codeql/cpp-queries@0.9.0,microsoft/windows-drivers@1.0.13
)
if "%WHCP_LEVEL%"=="WHCP_24H2" (
  set CODEQL_PACK_DEL=codeql/cpp-queries@0.10.0,microsoft/windows-drivers@1.2.0
  set CODEQL_PACK_GET=codeql/cpp-queries@0.9.0,microsoft/windows-drivers@1.1.0
)
if "%WHCP_LEVEL%"=="WHCP_25H2" (
  set CODEQL_PACK_DEL=
  set CODEQL_PACK_GET=codeql/cpp-queries@0.10.0,microsoft/windows-drivers@1.2.0
)
if "%WHCP_LEVEL%"=="WHCP_NEXT" (
  rem USE THIS AS A TEMPLATE FOR FUTURE VERSIONS
  rem NOTE: Remember to populate the CODEQL_PACK_DEL variable for all WHCP_LEVELS above with __NEXTRANGE__ packs [per WHCP_LEGACY example]...
  set CODEQL_PACK_DEL=
  set CODEQL_PACK_GET=codeql/cpp-queries@__NEXTRANGE__,microsoft/windows-drivers@__NEXTRANGE__
)
```
This would allow co-existence in the future regardless of whether or not the WHCP chooses to accept DVLs built with WHCP_25H2 for WHCP_24H2 or earlier targets. It also allows you to adjust rapidly should MS change their position.

Similarly, to pick a CodeQL test suite, specify in the `CODEQL_DRIVER_SUITES_REPO_HASH` environment variable, the relevant SHA256 Git commit from the `microsoft/Windows-Driver-Developer-Supplemental-Tools` repo.

Architecturally, I'm curious as to whether you think it worthwhile to split this framework, or the whole CodeQL part off to separate file(s). The natural extension of this line of thinking would be to consider whether it is worthwhile to split the whole _SDV build process off to separate file(s) - and maybe rename to _ANALYSIS or similar in the process. This is probably worth considering now if I'm splitting up this PR...

Best regards,
Ben

benyamin-codez commented:
I've pushed those changes:

Addendum to https://github.com/virtio-win/kvm-guest-drivers-windows/pull/1212/commits/da96c8982427f3e9137d74c7cf532d398589ec3c.

1. Further changes to `buildAll.bat`:
   a)  Added "@" prefix to `FOR` do commands
   b)  Extended logging around creation of COMPAT (RS1/1607) DVL files
   c)  Added SDV and Legacy DVL (DVL files created with legacy EWDKs) reporting. The latter queries the `<PROJECT>.legacy_dvl_result.txt` log file created by `build\makeLegacyDVLs.bat` and reports accordingly. Please see PR [#1210](https://github.com/virtio-win/kvm-guest-drivers-windows/pull/1210) for more information.
   d)  Consolidates exiting function (whether success or fail) to `:leave`
   e)  Creates a scheduled task to call `build\clean_build_log.bat` before leaving.

2. Created `build\clean_build_log.bat` to remove colour pallette artefacts from `build_log.txt` created during HCK/CI using Jenkins config. This script also deletes the scheduled task created in 1(e) above.

@kostyanf14

That log clean in (2) above, should complete while signing the drivers. There are presently 8 seconds of `timeout`, 3s to wait for handles to close and 5s for devs to read results in `stdout`. It usually completes within 10s. Signing a full tree of drivers usually takes longer than this in my environments. That being said, you might want to consider adding a short delay (perhaps using `timeout` or `Powershell Start-Sleep`) between the driver signing and the archiving of the artefacts for submission to ensure the `build_log.txt` is appropriately prepared prior to archiving.

benyamin-codez commented:
@YanVugenfirer @kostyanf14 @ybendito @vrozenfe

I have removed the packaging folder from `root\buildAll.bat` per PR #1209 and the folder itself per PR #1224.

Any final thoughts before I begin the process of splitting this up as discussed above...?

kostyanf14 commented:
@benyamin-codez 

>Any final thoughts before I begin the process of splitting this up as discussed above...?

I think you can split patches for now.

benyamin-codez commented:
@YanVugenfirer @kostyanf14 @ybendito @vrozenfe

I've done the first round of splits...

PRs
#1231
#1232 
#1233 
#1234 

Please excuse seasonal ASCII art in PR #1231.
Hope you are all enjoying the annual festivities despite local happenings - to the extent that is possible for you to do so.
Cheers,
Ben 8^D

kostyanf14 commented:
@benyamin-codez 

Hi,

We merged all the following PRs
https://github.com/virtio-win/kvm-guest-drivers-windows/pull/1231
https://github.com/virtio-win/kvm-guest-drivers-windows/pull/1232
https://github.com/virtio-win/kvm-guest-drivers-windows/pull/1233
https://github.com/virtio-win/kvm-guest-drivers-windows/pull/1234

Can we close this PR or there is still some fixes?

benyamin-codez commented:
@kostyanf14 

> We merged all the following PRs ... Can we close this PR or there is still some fixes?

Thanks Kostiantyn..!

Methinks we still have two more PRs.
I'll try to raise them ASAP, reference here and close this one.
They are extensions not fixes.

The two I had in mind were:

1. CodeQL Environment Prep (this is where I started after all)
2. Colour palette for trace messages

Regarding [1], to continue the trend, I'll refactor this as an external batch script. I am minded to also include the CodeQL run in the external batch script too, and name it `build\CodeQL_bld.bat`. One potential issue I saw here was related to whether or not your HCK-CI builds have `git` installed (otherwise the `CODEQL_RUN_BLIND` environment variable needs to be set) and if they do have `git`, whether or not they have internet access to download suites and CodeQL packages (otherwise the `CODEQL_OFFLINE_ONLY` environment variable needs to be set). If either of these environment variables would need to be set, and that is a problem for you, that would be helpful to know.

As for [2], you may recall I had proposed implementing a scheduled task that runs once the `buildAll.bat` script exits. This is used to clean the colour palette artefacts from any `build_log.txt` file - if it exists. This is necessary because until the batch script closes, the handles to the log file will not close so that it can then be opened for editing.

iirc, in your HCK-CI environment you only call the `buildAll.bat` script in the root directory and without any arguments. If so, this would likely mean no changes are required to integrate the mechanism into existing workflows whilst also permitting colourised output when building from the CLI or with the menu (per PR #1213).

Again, if any of this is a problem, it would be helpful to know now. That being said, the alternative would be to explicitly enable the feature via an environment variable, and this would be tedious, very messy and nowhere near as cool. 8^D

<details>
<summary>I do note, my previous comments regarding the delay between the build script running and the collection of any build artefacts, reproduced <ins>under this dropdown</ins> for your convenience.</summary>
<br>

> That log clean in (2) above, should complete while signing the drivers. There are presently 8 seconds of `timeout`, 3s to wait for handles to close and 5s for devs to read results in `stdout`. It usually completes within 10s. Signing a full tree of drivers usually takes longer than this in my environments. That being said, you might want to consider adding a short delay (perhaps using `timeout` or `Powershell Start-Sleep`) between the driver signing and the archiving of the artefacts for submission to ensure the `build_log.txt` is appropriately prepared prior to archiving.
</details>

I would like to highlight that it is unlikely you would need to implement any delay, as the handle to the log file would likely be closed within 4 to 5 seconds. The scheduled task would be deleted approximately 5 seconds later.

Did you have any thoughts about the above you would like to share before I start the split?
In this way I can accommodate your feedback in the next cut.

Also, did you want me to raise the PRs with `[CI-NO-BUILD]`, at least to start with?

Best regards,
Ben

kostyanf14 commented:
@benyamin-codez 

>Also, did you want me to raise the PRs with [CI-NO-BUILD], at least to start with?

Yes

benyamin-codez commented:
@YanVugenfirer @kostyanf14 @ybendito @vrozenfe

Ok, here is a summary of the next (last) round of splits.

First we have some `buildAll.bat` tidy ups:

PR #1282
PR #1283

Then we add colour:

PR #1284 (will need to be rebased - contains all previous commits)

Then we have:

PR #1285 (for `buildAll.bat` - will need to be rebased)
PR #1286 (for `build\build.bat` + new - will need to be rebased)

And finally:

PR #1287 (will need to be rebased - contains all other commits)

This should finalise this PR so I will close it.
Needed wiki updates are excluded.
8^d

Best regards,
Ben
